### DISCUSSION TITLE
TEST: LOAD "tpi:filename.dck"

### DISCUSSION THREAD

**Discussion by factus10** [timex-sinclair-projects/TS-Pico#20] [General]
**Test the ability to LOAD DCK files into the SRAM and Flash memory.**

First, [download the entire archive to your computer using this link](https://github.com/timex-sinclair-projects/TS-Pico/archive/refs/heads/Beta.zip). This will save a ZIP file to your download folder or where ever you specify.

Open the ZIP file and navigate to the **files** directory and expand the **DCKs and ROMs.zip** file.

Copy several DCK files to your SD card. Return the SD card to your TS-Pico.

### Testing the command sequence
With the SD cards in your TS-Pico, turn on your 2068. Use SAVE "tpi:cd _directory_" to navigate to the directory that contains your DCK files. Then, enter the following commands:
```
LOAD "tpi:filename.dck"
LOAD ""
```

The first command readies the DCK file. The TS-Pico uses the file extension to decide what to do with the file:

- TAP files are mounted as if you were inserting a cassette tape into a virtual tape player. 
- DCK files will be "inserted" into DOCK memory, either in the Static RAM or Flash ROM.

The next command, **LOAD ""**, causes the TS-Pico to send a special program to the 2068. The program will ask you a few questions about whether to save the DCK image on Static RAM or Flash and which of the available slots.

If you indicate that you want to keep the configuration, the Pico will reset and use the image you've loaded as a mounted DCK.

Otherwise you'll need to issue _SAVE "tpi:memdock" CODE nn,mm_ to activate the virtual cartridge, with _nn=1_ to use the SRAM, _nn=2_ use the Flash, and _mm_ is an even number between 0 and 14 to select the slot. 

Any other values will result in error. 

Also, notice currently the TS-Pico doesn't validate whether there was a previous DCK image in the selected slot; actually, the selected slot gets wiped out prior to writing the image. Use with caution. 

As a security measure, you cannot write anything (DCK or ROM) on slots 0 and 1 on the Flash. It is advisable to use SRAM whenever possible; Flash should only be used to store images between reboots.

From this moment on, the image will act exactly as if a physical cartridge is inserted on the slot. Some cartridges must be activated via OUT 244, 3; some others just need a NEW command, and even some others need a physical reset to start.

#### Recommended tests

- Test Androids.DCK, especially in the joystick mode. Are you able to play the game with a joystick?
- Test any other DCK file that uses a joystick. Are you able to play that game with the joystick?

**Comment by BetaTesterK** [comment]
I'll be able to get back to testing this weekend.  I'll look into this test specifically then.

**Comment by factus10** [comment]
Adam reported:

> I am having trouble mounting cartridges, but I'm also having trouble simply changing directories. This is with me using the most-up-to-date files when I restored my TS-Pico a couple of days ago. Now that I have access to GitHub again, perhaps this post belongs in the ticket thread for the Change Directory issues. For now, it's here on the discussion forum.
> 
> For instance, see the attached screenshot.
> 
> In the picture you'll see a directory listing of my TAP directory. Verbose is turned on. I tried to change to the DCK directory and I got the "Invalid I/O Command." Notice that verbose message tells me, "OS error changing to: DCK." I double-checked that the command was correct when I typed it.
> 
> With errors like this one, I begin to question my firmware again. Before I issued this CD command, I had just checked my microSD card in my Windows computer using chkdsk. There were no problems found. I did NOT reformat the microSD drive and move the files to it again. I have to trust chkdsk or how can I trust any of the files on my microSD card?
> 
> At a point like this one, I really do want to get the debug function working with Thonny and my TS-Pico. I believe that is the only way I'll be able to give usable feedback in a case like this one where my error reporting is vague.
> 
> Update:
> 
> When I began typing this message, I had not turned off the TS-Pico, nor had not rebooted the T/S 2068 after the directory error that I talk about above. This is good, as it leads me to another odd quirk that I can't explain.
> 
> After I typed this post, but before I sent it, I noticed that the fast-blinking on my TS-Pico had stopped. No LED was on at all. I'm not sure when that happened. It must have been about ten minutes after the above error occurred. Out of curiosity, I got a directory listing. I did not expect the TS-Pico to respond correctly. Instead, I got a directory listing, but not of the TAP directory. I got a directory with these files:
> 
> Notice that I did not change above the TAP directory to the "real" root directory. The last command I entered that gave me the error was meant to bring me into the DCK directory. After I issued the directory command that gave the above listing that I typed, the LED went from not blinking at all to the steady "heartbeat" blinking.
> 
> This seems weird, right?
> 
> When I come back into town this coming Sunday (Aug. 25), I'll still keep notes of what I do as I beta test the TS-Pico the following week, but I think I may also begin taking video of the TS-Pico and my TV for self-reference if I can manage to come up with an easy way to do it each time that I'm using the TS-Pico. It may help me (and others?) troubleshoot problems that I run into that are hard to track down using just my descriptions.

![359314393-c184946a-0423-4785-af26-79e9acb2b350](https://github.com/user-attachments/assets/93aeb778-9b75-4a32-ba7f-d6cad72c8699)

**Comment by factus10** [comment]
Adam reports:

> I tried to load eToolkit, the first cartridge on the list, and my TS-Pico crashed with an Invalid I/O device. I got the fast-blinking LED. I turned off my T/S 2068 and tried again with the same result.
> 
> Also, subtle typing errors crash the TS-Pico. For instance, this command changes directories correctly:
> 
> SAVE "TPI:CD DCK"
> 
> This mistyped command crashes the TS-Pico:
> 
> LOAD "TPI:CD DCK"
> 
> Yes, that is an error, but it's an error that's VERY easy to make. This causes a "never-ending, rapid blinking" LED. From this simple mistake, I am forced to reset my computer.
> 
> Once I get fast-blinking errors, I start to doubt that the data on my SD card is valid. For instance, is that why the eToolkit.dck doesn't load for me? Do I now have bad data? It's extremely hard to tell without putting the data on the SD card again.

**Comment by factus10** [comment]
Ricardo reports:

> Moctezuma's revenge shows the same behaviour in my system; both saving to Flash and SRAM. Testing the rest of the larger than 8Kb images, my results are :
>  
> eToolkit : no Flash; ok SRAM
> Flight Sim: no Flash; ok SRAM
> Manic Miner: ok Flash; ok SRAM
> Penetrator: ok Flash; ok SRAM
> Pinball: ok Flash; ok SRAM
> States&Cap: no Flash; ok SRAM
> Swordfight: ok Flash; ok SRAM
> Time Gate: ok Flash; ok SRAM
>  
> Notably:
> 1) All of the DCK images run from SRAM; but not all from Flash
> 2) The DCKs that fail, show about the same screen garbage
>  
> This leads me to think there's something subtle that's preventing the DCKs to run from Flash; maybe something related to the writing process. Perhaps somebody else can do the testing to see exactly what's going wrong with that. For example, try to burn those images on a real cartridge EPROM and see if they work; or comparing the image stored in Flash with the one supposed to be
>  
> What did you experience?

**Comment by factus10** [comment]
Adam reports:

> I successfully loaded the "Montzuma's Revenge" DCK file with the TS-Pico.  I get to the title screen, then when the game begins the sprites don't appear on the first screen and I can't figure out any way to continue, start the game or back out to the title screen again.
> 
> I checked the game's DCK image and loaded it with the FUSE emulator in T/S 2068 mode and the game begins fine and I'm able to play it (poorly) using the keyboard with keys 1-5 for movement.
> 
> What size DCK files does the TS-Pico support?  I've only had luck with 8K files so far, but my testing is minimal at best.
> 
> Can someone else test this "Montzuma's Revenge" DCK with their TS-Pico, please?
